### PR TITLE
Add launch handler article

### PR DESCRIPTION
**Pull request by tomayac**
Fixes https://github.com/GoogleChrome/web.dev/issues/6866
Changes proposed in this pull request:

- Add the article

### CODE REVIEW

File: src/site/content/en/blog/declarative-link-capturing/index.md [hunk 1]
```diff
@@ -44,13 +43,13 @@ separate app windows, if the app is open already, should a new window or tab be
...
-* Navigations caused by [file handlers](/file-handling/)
-* Navigations caused by the [Share Target API](/web-share-target/)
-* …and others
+- Clicked links from other web pages.
+- URL launches from a platform-specific app in the operating system.
+- Navigations originating from the [App Shortcuts API](/app-shortcuts/)
+- Links that go through [URL protocol handlers](/url-protocol-handler/)
+- Navigations caused by [file handlers](/file-handling/)
+- Navigations caused by the [Share Target API](/web-share-target/)
+- …and others
```
Comment (jpmedley) [inline src/site/content/en/blog/declarative-link-capturing/index.md:52]:
```suggestion
- Navigations originating from the [App Shortcuts API](/app-shortcuts/).
- Links that go through [URL protocol handlers](/url-protocol-handler/).
- Navigations caused by [file handlers](/file-handling/).
- Navigations caused by the [Share Target API](/web-share-target/).
- …and others.
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+{% Aside %} The Launch Handler API is part of the [capabilities project](/fugu-status/) and is
+currently in development. This post will be updated as the implementation progresses. {% endAside %}
+
+## What is the Launch Handler API? {: #what }
+
+There are many ways to launch a Progressive Web App. Probably the most common is via the icon on the
+home screen or the app drawer of the device. But when you think about it, there are many other ways
+a launch can happen:
+
+- The app can be launched as the result of a share action when the app is a
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:35]:
```suggestion
- A share action when the app is a
```

Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:35]: This is one of the permissible uses of a sentence fragment.

File: src/site/content/en/blog/launch-handler/index.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+  [share target](/web-share-target/).
+- A user may have clicked a file in their file explorer and decided to open it with the PWA, which
+  acts as a [file handler](/file-handling/).
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:38]:
```suggestion
- A user click in the file explorer opening the PWA, which
  acts as a [file handler](/file-handling/).
```

Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:38]: Lists should maintain a parallel grammatical structure.

File: src/site/content/en/blog/launch-handler/index.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+- If the PWA has registered as a [protocol handler](/protocol-handling/), the PWA may launch as the
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:39]:
```suggestion
- A registered [protocol handler](/protocol-handling/), that the PWA may launches as the
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  result of a matching protocol.
+- The PWA may also be launched as a result of a click on a
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:41]:
```suggestion
- A click on a
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+  [push notification](/push-notifications-overview/) or an [app icon shortcut](/app-shortcuts/).
+
+There are even more ways, but you get the idea… Given the manyfold possibilities for launching PWAs,
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:44]:
```suggestion
There are even more ways, but you get the idea. Given the manifold possibilities for launching PWAs,
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+what was missing so far is a way to let apps customize their launch behavior across all types of app
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:45]:
```suggestion
what has been missing is a way to let apps customize their launch behavior across all types of app
```

Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:45]: This verb tense is called 'past perfect' if you decided to look it up.

File: src/site/content/en/blog/launch-handler/index.md [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+}
+```
+
+### The `launch_handler` manifest member
+
+To declaratively specify the launch behavior of your app, add the `launch_handler` manifest member
+to your manifest. It has two sub-fields, `route_to` and `navigate_existing_client`. The former lets
+you control whether a new or an existing client should be launched, and the latter how and if this
+client should be navigated. The code snippet below shows an excerpt of a Web App Manifest file with
+exemplary values that would always route all launches to a new client.
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:108]:
```suggestion
client should be navigated. The Web App Manifest excerpt below shows a file with
exemplary values that would always route all launches to a new client.
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+If unspecified, `launch_handler` defaults to
+`{"route_to": "auto", "navigate_existing_client": "always"}`. The allowed values for the sub-fields
+are as follows:
+
+- `route_to`:
+
+  - `new-client`: A new browsing context is created in a web app window to load the launch's target
+    URL.
+  - `existing-client`: The most recently interacted with browsing context in a web app window for
+    the app being launched is chosen to handle the launch. How the launch is handled within that
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:128]:
```suggestion
  - `existing-client`: An existing browsing context is used, specifically the one most recently used. How the launch is handled within that
```

Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:128]: The exiting sentence was a bit of a mouthful. If I didn't get this right, we'll need a third option.

File: src/site/content/en/blog/launch-handler/index.md [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    browsing context depends on `navigate_existing_client`.
+  - `auto`: The behavior is up to the user agent to decide what works best for the platform. For
+    example, mobile devices only support single clients and would use `existing-client`, while
+    desktop devices support multiple windows and would use `new-client` to avoid data loss.
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:132]: So I should use `auto` unless I'm doing something specifically for desktop or mobile? Should we say that?

So `existing_client` can cause data loss. Shouldn't you say that in the description of that option?

File: src/site/content/en/blog/launch-handler/index.md [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+  - `always`: Existing browsing contexts chosen for launch will navigate the browsing context to the
+    launch's target URL.
+  - `never`: Existing browsing contexts chosen for launch will not be navigated and instead have
+    `targetURL` in the enqueued `LaunchParams` set to the launch's target URL.
+
+Both `route_to` and `navigate_existing_client` also accept a list (array) of values, where the first
+valid value will be used. This is to allow new values to be added to the spec without breaking
+backwards compatibility with existing implementations.
+
+For example, if the value `"matching-url-client"` were added, sites would specify
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:144]:
```suggestion
For example, if the hypothetical value `"matching-url-client"` were added, sites would specify
```

Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:144]: I don't like adding adjectives, but I wanted to make it clearer that this is not a promise to add this value.

File: src/site/content/en/blog/launch-handler/index.md [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+### The `window.launchQueue` interface
+
+If the app has declared that it wants to handle launches in an existing client (by specifying
+`"route_to": "existing-client"`), it can imperatively do something with incoming launch URLs.
+This is where the `launchQueue` comes into play. To access launch target URLs, a site needs to
+specify a consumer for the `window.launchQueue` object, which is then passed the target URL via the
+`launchParams.targetURL` field. Launches are queued until they are handled by the specified
+consumer, which is invoked exactly once for each launch. In this manner, every launch is handled,
+regardless of when the consumer was specified. The code snippet below shows a fictive audio player
+PWA that extract a song ID from a target URL that it is potentially passed upon launch.
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:157]:
```suggestion
PWA that extracts a song ID from a target URL that it is passed on launch.
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+<!--lint disable no-literal-urls -->
+1. Install the app on a Chrome OS device.
+1. Send yourself a link in a chat application of the form
+   <code>https://launch-handler.glitch.me?track=<strong>https://example.com/music.mp3</strong></code>.
+   (You can customize
+   <code><strong>https://example.com/music.mp3</strong></code> for any URL pointing to an audio file, for example,
+   <a href="https://launch-handler.glitch.me?track=https://cdn.glitch.me/3e952c9c-4d6d-4de4-9873-23cf976b422e%2Ffile_example_MP3_700KB.mp3?v=1638795977190"><code>https://launch-handler.glitch.me?track=https://cdn.glitch.me/3e952c9c-4d6d-4de4-9873-23cf976b422e%2Ffile_example_MP3_700KB.mp3?v=1638795977190</code></a>).
+1. Click the link in your chat app and notice how _Musicr 2.0_ opens and plays the track.
+1. Click the link in your chat app again and notice how always just one instance of Musicr 2.0 exists.
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:182]:
```suggestion
1. Click the link in your chat app again and notice that you will not get a second instance of Musicr 2.0.
```

File: src/site/content/en/blog/launch-handler/index.md [hunk 14]
```diff
@@ -0,0 +1,244 @@
...
+<!--lint enable no-literal-urls -->
+
+{% Img src="image/8WbTDNrhLsU0El80frMBGE4eMCD3/nFgE69N5VTYBDon1DGmB.png", alt="Musicr 2.0 web application launched and playing music it handled from the launch params.", width="800", height="235" %}
+
+## Security and permissions
+
+The Chromium team has designed and implemented the Launch Handler API using the core principles
```
Comment (jpmedley) [inline src/site/content/en/blog/launch-handler/index.md:189]:
```suggestion
The Chromium team designed and implemented the Launch Handler API using the core principles
```

**Review by jpmedley — CHANGES_REQUESTED** [review]

File: src/site/content/en/blog/launch-handler/index.md (hunk 14 above)
Comment (jpmedley) [reply src/site/content/en/blog/launch-handler/index.md:189]: I should have caught this a long time ago. I've [opened a PR](https://github.com/GoogleChrome/web.dev/pull/7033) to fix this elsewhere.

File: src/site/content/en/blog/launch-handler/index.md [hunk 15]
```diff
@@ -0,0 +1,244 @@
...
+```
+
+## Demo
+
+You can see a demo of the Launch Handler API in action in the [PWA Launch Handler Demo][demo]. Be
+sure to check out the [source code][demo-source] of the application to see how it uses the Launch
+Handler API.
+
+<!--lint disable no-literal-urls -->
+1. Install the app on a Chrome OS device.
```
Comment (tomayac) [inline src/site/content/en/blog/launch-handler/index.md:175]:
```suggestion
1. Install the _Musicr 2.0_ app on a Chrome OS device.
```

File: src/site/content/en/blog/launch-handler/index.md (hunk 10 above)
Comment (tomayac) [reply src/site/content/en/blog/launch-handler/index.md:131]: The `auto` value is the default, you get this when you don't specify anything else. I add a warning about data loss.

File: src/site/content/en/blog/launch-handler/index.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+  - `auto`: The behavior is up to the user agent to decide what works best for the platform. For
+    example, mobile devices only support single clients and would use `existing-client`, while
+    desktop devices support multiple windows and would use `new-client` to avoid data loss.
+
+- `navigate_existing_client`:
+  - `always`: Existing browsing contexts chosen for launch will navigate the browsing context to the
+    launch's target URL.
+  - `never`: Existing browsing contexts chosen for launch will not be navigated and instead have
+    `targetURL` in the enqueued `LaunchParams` set to the launch's target URL.
+
```
Comment (tomayac) [inline src/site/content/en/blog/launch-handler/index.md:138]:
```suggestion

{% Aside 'warning' %}
  Careful, setting `navigate_existing_client` to `always` can cause data loss. Only use this
  when it is safe to do so (for example, in a music player), or make sure the user's data gets
  saved before navigating away from the previous location.
{% endAside%}

```

File: src/site/content/en/blog/declarative-link-capturing/index.md [hunk 17]
```diff
@@ -27,9 +27,8 @@ tags:
 {% Aside 'caution' %} Declarative Link Capturing was part of the
 [capabilities project](/fugu-status/).
 The engineering team has decided that Declarative Link Capturing will _not_
-launch with the current design. Interested parties
-can check out and provide early feedback on the likely replacement feature called
-[Web App Launch Handling](https://github.com/WICG/sw-launch/blob/main/launch_handler.md).
+launch in the current shape. Instead, the feature has been redesigned as
```
Comment (jpmedley) [inline src/site/content/en/blog/declarative-link-capturing/index.md:30]:
```suggestion
launch with its current design. Instead, the feature has been redesigned as
```

Comment (jpmedley) [reply src/site/content/en/blog/declarative-link-capturing/index.md:30]: @tomayac Apologies. I didn't release this was a PTAL. I started reviewing as though it were new and found this. We should probably fix it.

**Review by jpmedley — CHANGES_REQUESTED** [review]
Apologies. I didn't release this was a PTAL. I started reviewing as though it were new and found this. We should probably fix it.

**Review by jpmedley — APPROVED** [review]

**Pull request merged**